nomad/structs/structs.go

```diff
@@ -7632,6 +7634,39 @@ func (a *Allocation) CopySkipJob() *Allocation {
 	return a.copyImpl(false)
 }
 
+func (a *Allocation) Canonicalize() {
+	if a.AllocatedResources == nil && a.TaskResources != nil {
+		ar := AllocatedResources{}
+		ar.Tasks = toAllocatedResources(a.TaskResources)
+
+		if a.SharedResources != nil {
+			ar.Shared.DiskMB = int64(a.SharedResources.DiskMB)
+			ar.Shared.Networks = a.SharedResources.Networks.Copy()
+		}
+
+		a.AllocatedResources = &ar
+	}
+
+	// TODO: Investigate if we should canonicalize the job
+	// it may be out of sync with respect to the original job
+	// a.Job.Canonicalize()
+}
+
+func toAllocatedResources(taskResources map[string]*Resources) map[string]*AllocatedTaskResources {
```

Comment:
We may want to just inline this code into Canonicalize as it seems easy to get orphaned here if we are able to remove the canonicalization in the future.